Sources/src/FromSTMTestBeamData_source.cc

```diff
@@ -0,0 +1,335 @@
+
+
+  //================================================================
+  class STMTestBeamDataDetail : private boost::noncopyable {
+    std::string myModuleLabel_;
+    art::SourceHelper const& pm_;
+    unsigned runNumber_; // from file name
+    art::SubRunID lastSubRunID_;
+    std::set<art::SubRunID> seenSRIDs_;
+
+    unsigned currentFileNumber_;
+    std::string currentFileName_;
+    std::ifstream* currentFile_ = nullptr;
+
+    unsigned currentSubRunNumber_; // from file
+    unsigned currentEventNumber_;
+    unsigned maxEvents_;
+    int verbosityLevel_;
+    uint16_t channel_; // the channel (HPGe or LaBr) will be different for each file
+    int binaryFileVersion_;
+
+    int printAtEvent;
+
+    // A helper function used to manage the principals.
+    // This is boilerplate that does not change if you change the data products.
+    void managePrincipals ( int runNumber,
+        int subRunNumber,
+        int eventNumber,
+        art::RunPrincipal*&    outR,
+        art::SubRunPrincipal*& outSR,
+        art::EventPrincipal*&  outE);
+
+    public:
+    STMTestBeamDataDetail(const Parameters &conf,
+        art::ProductRegistryHelper &,
+        const art::SourceHelper &);
+
+    void readFile(std::string const& filename, art::FileBlock*& fb);
+
+    bool readNext(art::RunPrincipal* const& inR,
+        art::SubRunPrincipal* const& inSR,
+        art::RunPrincipal*& outR,
+        art::SubRunPrincipal*& outSR,
+        art::EventPrincipal*& outE);
+
+    void closeCurrentFile();
+  };
+
+  //----------------------------------------------------------------
+  STMTestBeamDataDetail::STMTestBeamDataDetail(const Parameters& conf,
+      art::ProductRegistryHelper& rh,
+      const art::SourceHelper& pm)
+    : myModuleLabel_("FromSTMTestBeamData")
+    , pm_(pm)
+    , runNumber_(0)
+    , currentFileNumber_(0)
+    , currentSubRunNumber_(-1U)
+    , currentEventNumber_(0)
+    , maxEvents_(conf().maxEvents())
+    , verbosityLevel_(conf().verbosityLevel())
+    , binaryFileVersion_(0)
+  {
+    rh.reconstitutes<mu2e::STMWaveformDigiCollection,art::InEvent>(myModuleLabel_, "HPGe");
+    rh.reconstitutes<mu2e::STMWaveformDigiCollection,art::InEvent>(myModuleLabel_, "LaBr");
+    rh.reconstitutes<mu2e::STMTestBeamEventInfo,art::InEvent>(myModuleLabel_);
+
+    currentSubRunNumber_ = 0;
+    currentEventNumber_ = 0;
+
+    printAtEvent = 0;
+
+  }
+
+
+  //----------------------------------------------------------------
+  void STMTestBeamDataDetail::readFile(const std::string& filename, art::FileBlock*& fb) {
+
+    // Open the input file
+    currentFileName_ = filename;
+    currentFile_ = new std::ifstream(currentFileName_.c_str(), std::ios::in | std::ios::binary);
+    if (!currentFile_->is_open()) {
+      throw cet::exception("FromSTMTestBeamData") << "A problem opening binary file " << currentFileName_ << std::endl;
+    }
+
+    // Extract the run number and subrun number from the file name
```

Comment:
The code to extract information from the file name is a great candidate for putting it in a small helper function or helper class.  That way the mainline of readFile reads cleanly.  If the helper classs/function is of general use, but it in Mu2eUtilities.  If it is of use only in this file, put it in an anonymous namespace at the top of the file.

Reply:
Done in [b453042](https://github.com/Mu2e/Offline/pull/856/commits/b453042068b1c4de70437c348d074f87485acb2d). This won't be generally useful since it is just the binary test beam filename that we need to decipher but I have put it in its own class in Sources